cloudinit/sources/DataSourceEc2.py

```diff
@@ -942,18 +949,105 @@ def convert_ec2_metadata_network_config(
             "match": {"macaddress": mac.lower()},
             "set-name": nic_name,
         }
+        # Configure policy-based routing on secondary NICs / secondary IPs to
+        # ensure outgoing packets are routed via the correct interface.
+        #
+        # If device-number is not present (AliYun or other ec2-like platforms),
+        # do not configure source-routing as we cannot determine which is the
+        # primary NIC.
+        if nic_metadata.get("device-number") and nic_idx > 1:
```

Comment:
One thing to be aware of when accessing IMDS in response to udev add events is that IMDS propagates asynchronously, and the data corresponding with a newly attached network interface may not be complete at the time that the handler is running.

Note also that device-number is not sufficient to uniquely identify an interface attachment.  Certain instance configuration may repeat device-number across multiple interfaces, for example as shown in https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/p5-efa.html

In those cases, you may also want to consult the `network-card` key, which represents the NetworkCardIndex API field.  It gets a bit confusing, though, because IMDS returns an http 404 error in multiple scenarios with different meanings:
* The interface attachment doesn't actually exist
* The interface attachment data hasn't yet propagated to IMDS (in which case you probably should retry)
* The returned value would be zero (which is a perfectly valid value and is seen on the instance's primary interface)

Reply:
Thanks for pointing this out, I have added an extra commit handling the non-uniqueness of `device-index`.